src/eth_provider/provider.rs

```diff
@@ -232,11 +232,33 @@ where
     }
 
     async fn transaction_by_hash(&self, hash: B256) -> EthProviderResult<Option<reth_rpc_types::Transaction>> {
-        Ok(self
-            .database
-            .get_one::<StoredTransaction>(into_filter("tx.hash", &hash, HASH_PADDING), None)
-            .await?
-            .map(Into::into))
+        let pipeline = vec![
+            doc! {
+                // Union with pending transactions with only specified hash
+                "$unionWith": {
+                    "coll": StoredPendingTransaction::collection_name(),
+                    "pipeline": [
+                        {
+                            "$match": {
+                                "tx.hash": format_hex(hash, HASH_PADDING)
+                            }
+                        }
+                    ]
+                },
+            },
+            // Only specified hash in the transactions collection
+            doc! {
+                "$match": {
+                    "tx.hash": format_hex(hash, HASH_PADDING)
+                }
+            },
+            // Only one document in the final result with priority to the final transactions collection if available
+            doc! {
+                "$limit": 1
+            },
```

Comment:
I couldn't find this in the docs, can you link where it says priority is given to the final transactions collection? All I found was `The order in which the combined result set documents are output is unspecified.` [here](https://www.mongodb.com/docs/manual/reference/operator/aggregation/unionWith/#mongodb-pipeline-pipe.-unionWith).

Reply:
Just fixed with a sort by block number to give priority to finalized transactions over pending ones (null is considered lower in priority):

https://www.mongodb.com/docs/manual/reference/operator/aggregation/sort/